lib/scan-includes.js

```diff
@@ -0,0 +1,110 @@
+const realFetchUrl = require("./fetch-url");
+const GITHUB_SERVICE = require("./services").GITHUB;
+
+
+/** Scans for recursively included paths from a root specification.
+ *
+ * @param {string} repositoryUrl The base URL of the repository. The function
+ * won't fetch anything outside of this scope. This must end with a `/`.
+ * @param {string} rootSpec A path relative to the `repositoryUrl` pointing to
+ * the root specification in the `specType` format.
+ * @param {"bikeshed"|"respec"} specType
+ *
+ * @returns {Promise<Set<string>>} Paths relative to `repositoryUrl` that are
+ * recursively included by the specification.
+ */
+module.exports = async function scanIncludes(repositoryUrl, rootUrl, specType, fetchUrl = realFetchUrl) {
+    if (!repositoryUrl.endsWith('/')) {
+        throw new Error(`${repositoryUrl} must end with a '/'.`);
+    }
+    let includeRE;
+    switch (specType) {
+        case "bikeshed":
+            includeRE = /^path:([^\n]+)$/gm;
+            break;
+        case "respec":
+            includeRE = /data-include="([^"]+)"|data-include='([^']+)'/g;
+            break;
+        default:
+            throw new Error(`Unexpected specification type: ${JSON.stringify(specType)}`);
+    }
+
+    /** All the files that successfully fetched go here. */
+    const recursiveIncludes = new Set();
+    /** When this gets back to empty, we're done searching includes. */
+    const inProgressFetches = new Set();
+
+    await new Promise((resolve, reject) => {
+        async function scanOneFile(path) {
+            try {
+                console.log("Scanning", path);
+                if (path.startsWith('/') ||
+                    path.includes('../') ||
+                    // Block schemes, a bit more aggressively than necessary.
+                    path.includes(':')) {
+                    return;
+                }
+                let resolvedUrl;
+                try {
+                    resolvedUrl = new URL(path, repositoryUrl).href;
+                    if (!resolvedUrl.startsWith(repositoryUrl)) {
+                        console.error(resolvedUrl, "doesn't start with", repositoryUrl);
+                        // Can't include paths outside the repository.
+                        return;
+                    }
+                } catch (e) {
+                    console.error(e);
+                    // Don't treat broken URLs as includes.
+                    return;
+                }
+
+                inProgressFetches.add(path);
+                recursiveIncludes.add(path);
+                try {
+                    console.log("Fetching", resolvedUrl);
+                    body = await fetchUrl(resolvedUrl, GITHUB_SERVICE);
+                    console.log("Got", resolvedUrl);
+                    for (const match of body.matchAll(includeRE)) {
+                        for (let i = 1; i < match.length; i++) {
+                            const included = match[i].trim();
```

Comment:
Nit: (to avoid explicit loop)
```suggestion
                        const included = match.slice(1).find(s => s && s.trim());
```

Reply:
It's even simpler with the 1-capture regex.